core/src/main/scala/cats/collections/TreeList.scala

```diff
@@ -0,0 +1,867 @@
+    }
+
+    def traverseTree[F[_]: Applicative, A, B, N <: Nat](ta: Tree[N, A], fn: A => F[B]): F[Tree[N, B]] =
+      ta match {
+        case Root(a) => fn(a).map(Root(_))
+        case Balanced(a, left, right) =>
+          (fn(a), traverseTree(left, fn), traverseTree(right, fn)).mapN { (b, l, r) =>
+            Balanced(b, l, r)
+          }
+      }
+
+    implicit def eqTree[A: Eq]: Eq[Tree[Nat, A]] =
+      new Eq[Tree[Nat, A]] {
+        val eqA: Eq[A] = Eq[A]
+        def eqv(l: Tree[Nat, A], r: Tree[Nat, A]): Boolean =
+          (l, r) match {
+            case (Root(a), Root(b)) => eqA.eqv(a, b)
+            case (Balanced(a, al, ar), Balanced(b, bl, br)) =>
+              eqA.eqv(a, b) && eqv(al, bl) && eqv(ar, br)
+            case _ => false
+          }
+      }
+
+    final class TreeListIterator[A](from: TreeList[A]) extends Iterator[A] {
+      private var nexts: List[Tree[Nat, A]] = from match {
+        case Trees(treeList) => treeList
+      }
+
+      def hasNext: Boolean = nexts.nonEmpty
+      def next(): A =
+        if (nexts.isEmpty) throw new NoSuchElementException("TreeList.toIterator exhausted")
+        else {
+          nexts.head match {
+            case Root(a) =>
+              nexts = nexts.tail
+              a
+            case Balanced(a, l, r) =>
+              nexts = l :: r :: nexts.tail
+              a
+          }
+        }
+    }
+
+    final class TreeListReverseIterator[A](from: TreeList[A]) extends Iterator[A] {
+      private var nexts: List[Tree[Nat, A]] = from match {
+        case Trees(treeList) => treeList.reverse
+      }
+
+      def hasNext: Boolean = nexts.nonEmpty
+
+      /*
+       * The cost to call next when left most item has depth D is
+       * D. We know that D <= log_2 N, so in the worst case iteration
+       * is N log N. A tree has most items at the deepest levels,
+       * to this implies that we do need O(log N) work in the average
+       * case to access a next item
+       *
+       * On the other hand, we tear down depth as we do this and save
+       * that state for the future, so it could be that the total work
+       * is still O(N). This is an open question.
+       */
+      @tailrec
+      final def next(): A =
+        if (nexts.isEmpty) throw new NoSuchElementException("TreeList.toReverseIterator exhausted")
+        else
+          nexts.head match {
+            case Root(a) =>
+              nexts = nexts.tail
+              a
+            case Balanced(a, l, r) =>
+              // switch the order
+              nexts = r :: l :: Root(a) :: nexts.tail
+              next()
+          }
+    }
+  }
+
+  import Impl._
+
+  private final case class Trees[A](treeList: List[Tree[Nat, A]]) extends TreeList[A] {
+    def prepend[A1 >: A](a1: A1): TreeList[A1] =
+      treeList match {
+        case h1 :: h2 :: rest =>
+          def go[N1 <: Nat, N2 <: Nat, A2 <: A](t1: Tree[N1, A2], t2: Tree[N2, A2]): TreeList[A1] =
```

Comment:
just curious, what's the purpose of having `go`? it's not recursive and is invoked only once right after definition

Reply:
I wanted to be able to name the types for better clarity: `N1` and `N2`. That's the only reason. I'll comment.